rust-symcrypt/src/rsa/mod.rs

```diff
@@ -0,0 +1,714 @@
+//!     17, 124, 191, 60, 163, 218, 149, 209, 207, 181,
+//! ];
+//!
+//! let pub_exp = [0, 0, 0, 0, 0, 1, 0, 1];
+//! let public_key = RsaPublicKey::set_public_key(&modulus, &pub_exp, RsaKeyUsage::Sign).unwrap();
+//!    
+//! // public_key can now be used via pkcs1, pss, or oaep.
+//!
+//! // get the size of the modulus and the size of the public exponent.
+//! let modulus_size = public_key.get_size_of_modulus();
+//! let pub_exp_size = public_key.get_size_of_public_exponent();
+//!
+//! ```
+//!
+use crate::errors::SymCryptError;
+use crate::NumberFormat;
+use std::ptr;
+
+pub mod oaep;
+pub mod pkcs1;
+pub mod pss;
+
+// InnerRsaKey is a wrapper around symcrypt_sys::PSYMCRYPT_RSAKEY.
+#[derive(Debug)]
+pub(crate) struct InnerRsaKey(pub(crate) symcrypt_sys::PSYMCRYPT_RSAKEY);
+
+// Must free inner key rather than the outer struct.
+impl Drop for InnerRsaKey {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptRsakeyFree(self.0);
+        }
+    }
+}
+
+/// Rsa Public and Private Key State.
+///
+/// [`RsaKeyPair`] represents an Rsa key pair and also contains the [`RsaKeyUsage`].
+#[derive(Debug)]
+pub struct RsaKeyPair {
+    inner: InnerRsaKey,
+    key_usage: RsaKeyUsage,
+}
+
+/// Rsa Public Key State.
+///
+/// [`RsaPublicKey`] represents an Rsa public key and also contains the [`RsaKeyUsage`].
+#[derive(Debug)]
+pub struct RsaPublicKey {
+    inner: InnerRsaKey,
+    key_usage: RsaKeyUsage,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+/// `RsaKeyUsage` will indicate if the [`RsaKeyPair`] or [`RsaPublicKey`] will be used for [`RsaKeyUsage::Sign`] or [`RsaKeyUsage::Encrypt`], or [`RsaKeyUsage::SignAndEncrypt`].
+/// This check is enforced by SymCrypt and will return [`SymCryptError::InvalidArgument`] if the key is used for the wrong purpose.
+pub enum RsaKeyUsage {
+    // When using [`RsaKeyUsage::Sign`], the intended usage for the key will be only Signing.
+    Sign,
+
+    /// When using [`RsaKeyUsage::Encrypt`], the intended usage for the key will be only Encryption.
+    Encrypt,
+
+    /// When using [`RsaKeyUsage::SignAndEncrypt`], the intended usage for the key can be either Signing or Encryption.
+    SignAndEncrypt,
+}
+
+// to_symcrypt_flag converts the RsaKeyUsage to the corresponding SymCrypt flag and only needed internally.
+impl RsaKeyUsage {
+    pub(crate) fn to_symcrypt_flag(&self) -> symcrypt_sys::UINT32 {
+        match self {
+            RsaKeyUsage::Sign => symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_SIGN,
+            RsaKeyUsage::Encrypt => symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_ENCRYPT,
+            RsaKeyUsage::SignAndEncrypt => {
+                symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_ENCRYPT | symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_SIGN
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
+/// [`RsaKeyPairExportBlob`] holds the values of the `[RsaKeyPair]` when the key pair is exported.  
```

Comment:
@NielsFerguson, Just as an FYI, we are exporting the `crt_coefficient` so that we can encode to ASN.1 [RFC](https://datatracker.ietf.org/doc/html/rfc8017#appendix-A.1.2)
```

     RSAPrivateKey ::= SEQUENCE {
         version           Version,
         modulus           INTEGER,  -- n
         publicExponent    INTEGER,  -- e
         privateExponent   INTEGER,  -- d
         prime1            INTEGER,  -- p
         prime2            INTEGER,  -- q
         exponent1         INTEGER,  -- d mod (p-1)
         exponent2         INTEGER,  -- d mod (q-1)
         coefficient       INTEGER,  -- (inverse of q) mod p
         otherPrimeInfos   OtherPrimeInfos OPTIONAL
     }
```